Fix race condition on saving CNI datastore file

mohi-kalantari opened the pull request:
When a pod is created, kubelet will call the CRI to setup the container and its networking and CRI might call the CNI in parallel multiple times.

The code in CNI had no locking mechanism for saving/reading the file and as a result, we could see some inconsistencies across cni datastore file.

This commit fixes the issue by adding locks to the files.
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Fixes race condition in CNI datastore by adding file locking and refactoring `UbiCNI` class for safer file operations.
> 
>   - **Behavior**:
>     - Adds file locking to `allocate_ips_for_pod` and `handle_del` in `ubi_cni.rb` to prevent race conditions when accessing the IPAM store.
>     - Refactors `handle_add` and `handle_del` to use `safe_write_to_file` for atomic file operations.
>     - Introduces `check_required_env_vars` and `validate_input_ranges` for input validation.
>   - **Refactoring**:
>     - Extracts logging logic to `log_environment`.
>     - Moves DNS setup to `setup_dns`.
>     - Refactors IP allocation logic into `allocate_ips_for_pod`.
>   - **Testing**:
>     - Updates `ubi_cni_spec.rb` to test new locking behavior and refactored methods.
>     - Adds tests for `check_required_env_vars` and `validate_input_ranges`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=ubicloud%2Fubicloud&utm_source=github&utm_medium=referral)<sup> for 2f9747bb368acfd5437ddd151f27a0ba72945594. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

aamederen commented on rhizome/kubernetes/lib/ubi_cni.rb:
```diff
@@ -20,14 +20,20 @@ def initialize(input_data, logger)
+    File.open(IPAM_STORE_FILE, "w") do |file|
+      file.flock(File::LOCK_EX)
+      file.write(JSON.pretty_generate(@ipam_store))
+    end
```
what about using `safe_write_to_file` here?

mohi-kalantari replied:
that can also cause a race condition. safe write first writes to a temp file then copies to the destination. we would still need the lock behaviour. although I don't know how that would turn out if we want to mix it with a lock.

jeremyevans replied:
File moves in the same file system are atomic, so writing to a temp file, and then moving it to the final location after written and closed should work fine.  I'd recommend that over the flock approach.

mohi-kalantari replied:
This approach doesn't work because file needs to remain locked when cni is allocating ips. now I've changed the code to fully protect that part

mohi-kalantari replied:
I used safe_write_to_file with the use of yielding blocks and it works fine. thanks.

aamederen commented on rhizome/kubernetes/lib/ubi_cni.rb:
```diff
@@ -20,14 +20,20 @@ def initialize(input_data, logger)
   def load_ipam_store
     if File.exist?(IPAM_STORE_FILE)
-      JSON.parse(File.read(IPAM_STORE_FILE))
+      File.open(IPAM_STORE_FILE, "r") do |file|
```
File::RDONLY instead of "r"?

mohi-kalantari replied:
r is more straightforward for reading string. File::RDONLY is used more with | operator which someone might want to combine flags with eachother.

jeremyevans replied:
`"r"` seems better than `File::RDONLY` here, but you could also omit the argument, as `"r"` is the default.

aamederen reviewed (approved):
Once you address the two comments, :shipit:

fdr commented on rhizome/kubernetes/lib/ubi_cni.rb:
```diff
@@ -49,189 +28,197 @@ def run
   end
 
+  def with_ipam_store
+    File.open(IPAM_STORE_FILE, File::RDWR | File::CREAT, 0o644) do |file|
```
ellipsis seems about 30% right here, is there a reason you are doing overwrites rather than atomic renames? It raises a host of partial completion issues.

mohi-kalantari replied:
I need to keep the lock during the ip allocation. that's why I'm not using the safe_write. If I read then manipulate and then save not like this, race condition can happen. you can look at allocate_ips_for_pod function to get a better sense.

fdr replied:
we discussed this a bit out of band, but basically, this is a well-trodden conflict in goals in what POSIX file systems support: the easiest method of atomic updates is the temp-file-flush-rename-flush-directory-entry, but because the inodes involved shuffle around in that process, advisory locking via flock doesn't work, since a file being named by the path is constantly changing.

 `.lock` (or even `.pid`, or any other stable sentinel file) as a spot to hold the lock is how this is traditionally solved, and why it shows up in so many unix programs.

mohi-kalantari commented:
will change to a separate lock file for the duration of ip allocation and then write using the safe_write.

fdr commented on rhizome/kubernetes/lib/ubi_cni.rb:
```diff
@@ -49,189 +28,199 @@ def run
+        ipam_store = {"allocated_ips" => {}}
+      end
+      result = yield ipam_store
+      safe_write_to_file(IPAM_STORE_FILE, JSON.pretty_generate(ipam_store))
```
I'm a bit confused about why with_ipam_store has to flock but also safe_write_to_file, perhaps safe_write_to_file needs to get a little smarter or refactored so other stuff can be done while it holds its lock. the code seems nearly in duplicate, but I suppose some kind of compositional weakness of safe_write_to_file is why you're doing that?

fdr commented on rhizome/kubernetes/lib/ubi_cni.rb:
```diff
@@ -49,189 +28,199 @@ def run
+  def setup_dns(cni_netns)
+    FileUtils.mkdir_p("/etc/netns/#{cni_netns}")
+    dns_config = <<~EOF
+nameserver 10.96.0.10
```
@ellipsis-div we don't want to take this advisory, stop

fdr commented on rhizome/common/lib/util.rb:
```diff
@@ -60,18 +60,33 @@ def sync_parent_dir(f)
+    begin
+      tmp_path = "#{path}.tmp"
+      File.open(tmp_path, "w") do |new_file|
+        result = yield(old_file, new_file)
```
should probably relieve the caller of having to do the data fsync and do it for them, before the File.rename

mohi-kalantari replied:
fixed

fdr commented on rhizome/common/lib/util.rb:
```diff
@@ -60,18 +60,33 @@ def sync_parent_dir(f)
+      File.open(tmp_path, "w") do |new_file|
+        result = yield(old_file, new_file)
+      end
+      File.rename(tmp_path, path)
```
after the rename, you probably want to then block on syncing the directory metadata, there's a helper for that already: `sync_parent_dir`

jeremyevans commented on rhizome/common/lib/util.rb:
```diff
@@ -60,18 +60,36 @@ def sync_parent_dir(f)
+end
+
+def atomic_file_replace(path)
+  lock_path = "/tmp/#{Digest::SHA256.hexdigest(path)}.lock"
```
Not that we expect foul play here, but this isn't a secure approach to creating a temporary file against an attacker with access to the file system, because if you know the path, you can calculate the SHA256 and setup a symlink to a file they control.  You can to use `File::EXCL` in the flags below, so that the code raises if the file already exists.  That would probably remove the need for the `flock`.  You would then need to handle exceptions raised if the file exists, so that concurrent processes wait until this process finishes.  I think that seems like a better approach than using `flock`.

I don't have time to review the other parts of this for at least a week, as I'm about to board a flight to Japan.

fdr replied:
I don't think that's necessary for this patch to do though, since this patch has application goals in mind, and was following with the current implementation for the lock file. We can reconsider the lock file separately

fdr commented:
Personally, I find this patch unreviewable: there’s a lot of churn unrelated to the stated purpose (a relatively minor change to use safe_write_to_file with a block to protect against race conditions and torn writes).

My standard for review is this: for any arbitrary line in any hunk, I ask myself, “Why is this here, with respect to the stated purpose in the commit message?” I need a good answer, or at least to be satisfied that I can’t reach one (for example, in an unfamiliar domain, like organizing web page templates, which is not my area of expertise).

To help you get started, I’ve separated out some patches. The commit messages are poor and should be improved, since I don’t know your detailed motivation for each change—I can only describe them mechanically. They may also need to be organized differently, but I lack the insight you had when writing them, so I’m just making educated guesses and splitting off what was relatively straightforward.

Even as-is, this reduction makes reading the final patch much easier, since interactions between lines in a patch grow quadratically:

text
fdr@albatross ~/c/clover ((4460b49d)) [1]> git diff HEAD^ --stat
 rhizome/kubernetes/lib/ubi_cni.rb       | 147 +++++++++++++++++-----------------
 rhizome/kubernetes/spec/ubi_cni_spec.rb | 453 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------
 2 files changed, 370 insertions(+), 230 deletions(-)
fdr@albatross ~/c/clover ((4460b49d))> git diff origin/main --stat
 rhizome/kubernetes/lib/ubi_cni.rb       | 246 ++++++++++++++++++++++++-----------------------
 rhizome/kubernetes/spec/ubi_cni_spec.rb | 549 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------
 2 files changed, 519 insertions(+), 276 deletions(-)

Still, there’s considerably more to be done. With full fluency in the patch and decent tool support, I suggest using git-cola—git add -p just isn’t sufficient for moderately complex cases.

fdr reviewed (changes requested):
with regrets, we gotta do something about the amount of stuff mixed in here.

fdr commented on rhizome/kubernetes/lib/ubi_cni.rb:
```diff
@@ -49,167 +28,182 @@ def run
+    safe_write_to_file(IPAM_STORE_FILE) do |file|
+      ipam_store = begin
+        content = File.read(IPAM_STORE_FILE)
+        content.empty? ? {"allocated_ips" => {}} : JSON.parse(content)
```
want to just crash and die here.

fdr commented:
I also have changes I'd like to make, but I'm not going to touch on those yet, because I'm still trying to deconstruct this patch to understand it. I may revisit to change the form of things. I notice there are downstream PRs merging into this one, so I don't want to touch the cumulative diffs. In the interest of project management, I'll do some revisitation for it later.

In general, here's the main thing I'm noticing:

There's way, way, way, way, way, way too much use of `allow`. As I am refactoring, it's hard to know if I'm activating the intended mocks. Use the unfortunately-wordy `expect(foo).to receive(:bar).at_least(:once)` if a method is being called multiple times...however, like it says, it checks the mock is being called (that is, used) *at least once*.

Related to this, some of the unit tests are written with involved `before` blocks that are kinda-sorta trying to make the tests both an integration test and a unit test, by mocking out the guts of a helper method when trying to write tests against functionality in the code that is calling the helper method. Do this much less: write ugly tests for difficult helper methods, and then mock the helper method simply when you are testing something else that does not have it as an emphasis.

fdr commented:
I also contemplated some of the really repetitive expect().to receive(:r).with()..

You could DRY this up, not so much in lines as the number of symbols/subexpressions with a proc...again, more diffs that I think can be done post-facto:

```diff
diff --git a/rhizome/kubernetes/spec/ubi_cni_spec.rb b/rhizome/kubernetes/spec/ubi_cni_spec.rb
index faa3968f..b737ed1d 100644
--- a/rhizome/kubernetes/spec/ubi_cni_spec.rb
+++ b/rhizome/kubernetes/spec/ubi_cni_spec.rb
@@ -98,25 +98,26 @@ def stub_safe_write_to_file(ipam_store)
       expect(ubicni).to receive(:check_required_env_vars).with(["CNI_CONTAINERID", "CNI_NETNS", "CNI_IFNAME"])
       expect(ubicni).to receive(:validate_input_ranges)
 
-      expect(ubicni).to receive(:r).with("ip link add veth_abcdef12 addr 00:aa:bb:cc:dd:ee type veth peer name eth0 addr 00:11:22:33:44:55 netns testnetns")
-      expect(ubicni).to receive(:r).with("ip -6 -n testnetns addr replace 2001:db8::2/128 dev eth0")
-      expect(ubicni).to receive(:r).with("ip -6 -n testnetns link set eth0 mtu 1400 up")
-      expect(ubicni).to receive(:r).with("ip -6 -n testnetns route replace default via fe80::02aa:bbff:fecc:ddee dev eth0")
-      expect(ubicni).to receive(:r).with("ip -6 link set veth_abcdef12 mtu 1400 up")
-      expect(ubicni).to receive(:r).with("ip -6 route replace 2001:db8::2/128 via fe80::0211:22ff:fe33:4455 dev veth_abcdef12 mtu 1400")
-
-      expect(ubicni).to receive(:r).with("ip -6 -n testnetns addr replace fd00::2/128 dev eth0")
-      expect(ubicni).to receive(:r).with("ip -6 -n testnetns link set eth0 mtu 1400 up")
-      expect(ubicni).to receive(:r).with("ip -6 link set veth_abcdef12 mtu 1400 up")
-      expect(ubicni).to receive(:r).with("ip -6 route replace fd00::2/128 via fe80::0211:22ff:fe33:4455 dev veth_abcdef12 mtu 1400")
-
-      expect(ubicni).to receive(:r).with("ip addr replace 192.168.1.1/24 dev veth_abcdef12")
-      expect(ubicni).to receive(:r).with("ip link set veth_abcdef12 mtu 1400 up")
-      expect(ubicni).to receive(:r).with("ip -n testnetns addr replace 192.168.1.100/24 dev eth0")
-      expect(ubicni).to receive(:r).with("ip -n testnetns link set eth0 mtu 1400 up")
-      expect(ubicni).to receive(:r).with("ip -n testnetns route replace default via 192.168.1.1")
-      expect(ubicni).to receive(:r).with("ip route replace 192.168.1.100/32 via 192.168.1.1 dev veth_abcdef12")
-      expect(ubicni).to receive(:r).with("echo 1 > /proc/sys/net/ipv4/conf/veth_abcdef12/proxy_arp")
+      expect_run = ->(cmd) { expect(ubicni).to receive(:r).with(cmd) }
+      expect_run["ip link add veth_abcdef12 addr 00:aa:bb:cc:dd:ee type veth peer name eth0 addr 00:11:22:33:44:55 netns testnetns"]
+      expect_run["ip -6 -n testnetns addr replace 2001:db8::2/128 dev eth0"]
+      expect_run["ip -6 -n testnetns link set eth0 mtu 1400 up"]
+      expect_run["ip -6 -n testnetns route replace default via fe80::02aa:bbff:fecc:ddee dev eth0"]
+      expect_run["ip -6 link set veth_abcdef12 mtu 1400 up"]
+      expect_run["ip -6 route replace 2001:db8::2/128 via fe80::0211:22ff:fe33:4455 dev veth_abcdef12 mtu 1400"]
+
+      expect_run["ip -6 -n testnetns addr replace fd00::2/128 dev eth0"]
+      expect_run["ip -6 -n testnetns link set eth0 mtu 1400 up"]
+      expect_run["ip -6 link set veth_abcdef12 mtu 1400 up"]
+      expect_run["ip -6 route replace fd00::2/128 via fe80::0211:22ff:fe33:4455 dev veth_abcdef12 mtu 1400"]
+
+      expect_run["ip addr replace 192.168.1.1/24 dev veth_abcdef12"]
+      expect_run["ip link set veth_abcdef12 mtu 1400 up"]
+      expect_run["ip -n testnetns addr replace 192.168.1.100/24 dev eth0"]
+      expect_run["ip -n testnetns link set eth0 mtu 1400 up"]
+      expect_run["ip -n testnetns route replace default via 192.168.1.1"]
+      expect_run["ip route replace 192.168.1.100/32 via 192.168.1.1 dev veth_abcdef12"]
+      expect_run["echo 1 > /proc/sys/net/ipv4/conf/veth_abcdef12/proxy_arp"]
 
       output = ubicni.handle_add
       response = JSON.parse(output)
```

The square brackets on a proc allow you to omit `.call(args)`, though that's a lesser known feature.

mohi-kalantari commented:
Aside from a specific test which called a function multiple times, I changed all `allow`s to `expect`s. Thanks for the suggestion

Also fixed the repetitive expect().to receive(:r).with() with the lamda function you wrote.

Also changed tests to not mock helpers using before blocks.

@fdr

fdr commented:
I have decomposed it and need to run it by you and maybe do some final surgeries on less important parts, but now I understand the most important parts rigorously. My force-push ends up cumulatively identical to what it was.

fdr commented:
Alright, the patch has been decomposed. The following changes  were left out of the cumulative difference

```diff
diff --git a/94632323 b/18e3f17c
index 94632323..18e3f17c 100644
--- a/94632323
+++ b/18e3f17c
@@ -48,6 +48,7 @@ class UbiCNI
     subnet_ula_ipv6 = @input_data["ranges"]["subnet_ula_ipv6"]
     subnet_ipv6 = @input_data["ranges"]["subnet_ipv6"]
     subnet_ipv4 = @input_data["ranges"]["subnet_ipv4"]
+
     container_id = ENV["CNI_CONTAINERID"]
     cni_netns = ENV["CNI_NETNS"].sub("/var/run/netns/", "")
     inner_ifname = ENV["CNI_IFNAME"]
@@ -59,15 +60,14 @@ class UbiCNI
     outer_link_local = mac_to_ipv6_link_local(outer_mac)
     outer_ifname = "veth_#{container_id[0, 8]}"
 
-    @logger.info "Setting up veth pair for container #{container_id}"
-    r "ip link add #{outer_ifname} addr #{outer_mac} type veth peer name #{inner_ifname} addr #{inner_mac} netns #{cni_netns}"
-
     @logger.info "Configuring DNS for network namespace #{cni_netns}"
     setup_dns(cni_netns)
 
+    @logger.info "Setting up veth pair for container #{container_id}"
+    r "ip link add #{outer_ifname} addr #{outer_mac} type veth peer name #{inner_ifname} addr #{inner_mac} netns #{cni_netns}"
+
     @logger.info "Allocating IP addresses for container #{container_id}"
     ipv4_container_ip, ipv4_gateway_ip, container_ula_ipv6, container_ipv6 = allocate_ips_for_pod(container_id, subnet_ula_ipv6, subnet_ipv6, subnet_ipv4)
-
     @logger.info "Setting up IPv6 configurations"
     setup_ipv6(container_ipv6, inner_link_local, outer_link_local, cni_netns, inner_ifname, outer_ifname, setup_default_route: true)
     setup_ipv6(container_ula_ipv6, inner_link_local, outer_link_local, cni_netns, inner_ifname, outer_ifname)
@@ -144,7 +144,10 @@ options ndots:5
   def setup_ipv6(container_ip, inner_link_local, outer_link_local, cni_netns, inner_ifname, outer_ifname, setup_default_route: false)
     r "ip -6 -n #{cni_netns} addr replace #{container_ip}/#{container_ip.prefix} dev #{inner_ifname}"
     r "ip -6 -n #{cni_netns} link set #{inner_ifname} mtu #{MTU} up"
-    r "ip -6 -n #{cni_netns} route replace default via #{outer_link_local} dev #{inner_ifname}" if setup_default_route
+    if setup_default_route
+      r "ip -6 -n #{cni_netns} route replace default via #{outer_link_local} dev #{inner_ifname}"
+    end
+
     r "ip -6 link set #{outer_ifname} mtu #{MTU} up"
     r "ip -6 route replace #{container_ip}/#{container_ip.prefix} via #{inner_link_local} dev #{outer_ifname} mtu #{MTU}"
   end
@@ -152,9 +155,11 @@ options ndots:5
   def setup_ipv4(container_ip, gateway_ip, cni_netns, inner_ifname, outer_ifname)
     r "ip addr replace #{gateway_ip}/24 dev #{outer_ifname}"
     r "ip link set #{outer_ifname} mtu #{MTU} up"
+
     r "ip -n #{cni_netns} addr replace #{container_ip}/24 dev #{inner_ifname}"
     r "ip -n #{cni_netns} link set #{inner_ifname} mtu #{MTU} up"
     r "ip -n #{cni_netns} route replace default via #{gateway_ip}"
+
     r "ip route replace #{container_ip}/#{container_ip.prefix} via #{gateway_ip} dev #{outer_ifname}"
     r "echo 1 > /proc/sys/net/ipv4/conf/#{outer_ifname}/proxy_arp"
   end
@@ -176,6 +181,7 @@ options ndots:5
       file.flush
       file.fsync
     end
+
     "{}"
   end
 
@@ -204,6 +210,7 @@ options ndots:5
       ips: [{address: container_ip, gateway: nil, interface: 0}],
       dns: {nameservers: dns_servers, search: search_domains, options: ["ndots:5"]}
     }
+
     JSON.generate(response)
   end
 
```
